geode-core/src/integrationTest/java/org/apache/geode/management/internal/security/JmxMultipleConnectionsTest.java

```diff
@@ -0,0 +1,92 @@
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.management.internal.security;
+
+import static org.apache.geode.distributed.ConfigurationProperties.SECURITY_MANAGER;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.management.remote.JMXConnector;
+import javax.management.remote.JMXConnectorFactory;
+import javax.management.remote.JMXServiceURL;
+
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import org.apache.geode.cache.RegionShortcut;
+import org.apache.geode.management.MemberMXBean;
+import org.apache.geode.security.TestSecurityManager;
+import org.apache.geode.test.junit.categories.SecurityTest;
+import org.apache.geode.test.junit.rules.ConnectionConfiguration;
+import org.apache.geode.test.junit.rules.MBeanServerConnectionRule;
+import org.apache.geode.test.junit.rules.ServerStarterRule;
+
+@Category({SecurityTest.class})
+public class JmxMultipleConnectionsTest {
+  private MemberMXBean bean;
+
+  @ClassRule
+  public static ServerStarterRule server = new ServerStarterRule().withJMXManager()
+      .withProperty(SECURITY_MANAGER, TestSecurityManager.class.getName())
+      .withProperty(TestSecurityManager.SECURITY_JSON,
+          "org/apache/geode/management/internal/security/cacheServer.json")
+      .withRegion(RegionShortcut.REPLICATE, "region1")
+      .withAutoStart();
+
+  @Rule
+  public MBeanServerConnectionRule connectionRule =
+      new MBeanServerConnectionRule(server::getJmxPort);
+
+  @Before
+  public void setUp() throws Exception {
+    bean = connectionRule.getProxyMXBean(MemberMXBean.class);
+  }
+
+  @Test
+  @ConnectionConfiguration(user = "super-user", password = "1234567")
+  public void testMultipleJMXConnection() throws Exception {
+
+    assertThat(bean.processCommand("locate entry --key=k1 --region=region1")).isNotNull();
+
+    JMXServiceURL url =
+        new JMXServiceURL("service:jmx:rmi:///jndi/rmi://:" + server.getJmxPort() + "/jmxrmi");
+    Map<String, String[]> env = new HashMap<>();
+    String user = "super-user";
+    String password = "1234567";
+    env.put(JMXConnector.CREDENTIALS, new String[] {user, password});
+    Map<String, String[]> env1 = new HashMap<>();
+    String user1 = "region1-user";
+    String password1 = "1234567";
+    env1.put(JMXConnector.CREDENTIALS, new String[] {user1, password1});
+
+    // create JMX connection with the same and different credentials
+    JMXConnector jmxConnector = JMXConnectorFactory.connect(url, env);
+    JMXConnector jmxConnector1 = JMXConnectorFactory.connect(url, env1);
+
+    assertThat(bean.processCommand("locate entry --key=k1 --region=region1")).isNotNull();
```

Comment:
when you do this, you are using the connection that's opened by connectionRule itself, thus effectively put "super-user" back to the thread context, so the the jmx close event would log out the super-user. That's why I am asking the use case behind this test, why you would open a connection, yet use another connection to do operation?

Reply:
When we close connection that is not using for doing operation it sends jmx close notification, and in case we do operation before that, the current thread will be this started with connectionRule and it will logout it. After this logout we tried again to do operation, but this user is logged out and we have that user in this thread is null and test will fail.
With this solution close connection will not log out user from current thread, it will logout user which connection is closed.
So I agree that we don't need here another connection to the same user as it will not be use for doing operation. Why I add this connection is to be visible that for the users with the same credentials new Subject will be created and logging out one of them will not logout both.

Reply:
logout only logs out the user in the current thread, i.e. only one of them. If after you opened up your second connection,  you get the memberMxBean using the 2nd connection, you would not have this problem.  This is the test that I wrote to demonstrate this. ("data" user has access to the region, but "dataReadAnother" doesn't). 
```
@Test
  @ConnectionConfiguration(user = "data", password = "data")
  public void testMultipleJMXConnection() throws Exception {

    System.out.println("first");
    String first = memberMbean.processCommand("locate entry --key=k1 --region=test");
    System.out.println("first result: " + first);
    assertThat(first).isNotNull();

    JMXServiceURL url =
        new JMXServiceURL("service:jmx:rmi:///jndi/rmi://:" + server.getJmxPort() + "/jmxrmi");
    Map<String, String[]> env1 = new HashMap<>();
    String user1 = "dataReadAnother";
    String password1 = "dataReadAnother";
    env1.put(JMXConnector.CREDENTIALS, new String[] {user1, password1});

    // create JMX connection with the same and different credentials
    JMXConnector jmxConnector1 = JMXConnectorFactory.connect(url, env1);

    System.out.println("second");
    MemberMXBean anotherBean = JMX
        .newMXBeanProxy(jmxConnector1.getMBeanServerConnection(), getObjectName(jmxConnector1, MemberMXBean.class, null), MemberMXBean.class);
    try {
      anotherBean.processCommand("locate entry --key=k1 --region=test");
    }
    catch (Exception e) {
      assertThat(e.getMessage()).contains("dataReadAnother not authorized for DATA:READ:test:k1");
    }

    // close another jmx connections
    jmxConnector1.close();

    System.out.println("third");
    String third = memberMbean.processCommand("locate entry --key=k1 --region=test");
    System.out.println("third result: " + third);
    assertThat(third).isNotNull();
  }
```
Basically, because login/logout/authenticate all work with user on the current thread, and it's not a common practice to mix users interchangeably in the same thread. From your test code, it shows that we don't support  code sequences like: open connection1 -> open connection2 -> use connection1 to do operation -> log out connection 2 in one thread, but we do support: open connection 1 -> open connection 2 -> use connection 2 to do operation -> log out connection 2. That's why I keep asking in what use case that the first code sequence is necessary except in a contrived test? I don't like further caching the subject in a map solution, if it's up to me, I would rather document that we do not support these kind of code sequence.

Reply:
We don't support code sequences like: open connection1 -> open connection2 -> use connection1 to do operation -> log out connection 2 in one thread, as when it received jmx close notification currentThread will be connection1(last used) and it will logout user from connection1, and after receiving new jmx close notification it will get the same currentThread, but now user will be null as it is already logged out. In that case connection2 will be closed but user from it will not be logged out. 
Apart from that, is it normal behavior that every user which is logged in should be able to logout? Or we need to limit it to use only one user?

Reply:
probably a better solution at the time 2nd user logs in, we should log out the 1st user. We do not support multiple users in one thread, that should be the rule of thumb to abide by. So in the log in method, if we detect there is already a current user, log it out first.

Reply:
Hi @jinmeiliao I think that we have some misunderstanding here. Let's wrap up all.
When new user is logged in it use it's own thread(login method first remove ThreadContext so every user will have their own thread). When user is logout it get Subject from currentThread.
 For example,
1. _user1_ will have _currentUser=**Subject1**_ and _user2_ will have _currentUser=**Subject2**_. At this point _currentUser_ from the _currentThread_ is _**Subject2**_. 
2. If we do some operation with _user1_, _currentUser_ will be **_Subject1_**. If we close _user2_ connection it will send _jmx close notification_ and call logout method. The problem is that in logout method it get _currentUser_ from _currentThread_ which is in this case **Subject1**(last operation was with _user1_). So it will call logout on this _currentUser_, which means that it will logout _user1_.
3. If we after that try to do some operation with _user1_(we didn't close this connection, we close _user2_ connection) it will use _currentUser_ from _currentThread_ which is **null** now. And we got following exception `NotAuthorizedException: null not authorized for CLUSTER:READ`

So the problem is when calling logout method it doesn't know which connection is closed, it only knows that one jmx connection is closed, and it will use Subject from the currentThread. With caching Subject in map we can send which user connection is closed and what is it's subject so it can be successfully logged out.
So the point is that ever user which is logged in should be logged out.

Reply:
@mkevo I understand the problem when you do code sequence like: open connection1 -> open connection2 -> use connection1 to do operation -> close connection 2, which leads to closing connection2 will leads to logout user from connection1, and thus, leave user from connection2 not logged out. What I am asking is: do we really want to to support this code sequence? Is there a use case for it?

Reply:
It seems logical for me if some user which is logged in should be also logged out. 
We will always get this exception if after connection2 is closed and logout user1, we try to do some operation with user1(at this point we expect that user2 is logged out). It will fail with `NotAuthorizedException: null not authorized for CLUSTER:READ` despite of connection1 is not closed.
If we connect with two user to JMX to get some info, and user1 is over between user2 logging in and doing an operation user1 is logged out and user2 try to do some operation it will fail.

Reply:
We use Shiro in the backend for security, and Shiro puts subject in the thread local so in essence, we can only support one user in one thread. You do have a good point that every user who logs in should log out, and in the login method of `IntegratedSecurityService`, we simply did a `ThreadContext.remove()` to wipe out existing user trace to make sure one thread should have one user only. What I had suggested before, is that in here, we should log out the existing user if any. This would ensure that we log out every user even if the developer has mixed users in one thread. By doing that, we protect ourselves to ensure that one thread can have only one active user. And this would work for not just jmx connection, but all other connections as well.

Reply:
Hi @jinmeiliao , sorry for a late response. 
We have JMX exporter connected to Geode and we are able to connect more than one Prometheus instances to Geode. So, in a case we create two or more connections in the same time, they will be connected to the same session. If one is logged out while the second doing some operation it will fail, and we will get this exception.

Reply:
you probably can do one of these two things: 1, do not use the same thread to open different connections to Geode on your client side, 2. rework your solution to support this scenario by somehow containing your logic finding the correct subject to logout at connection close time in "JMXShiroAuthenticator". Your current solution needs to add a `logoutJmxUser` to securtyService api which doesn't feel right, we don't have such specific method for other kind of user at all, so would like to keep those apis as general as possible.

Reply:
Thanks for the comments.
For me this first opinion is not a solution, it is just temporary workaround. About second opinion, do you have any idea how to refactor this as we need correct subject to logout. From my point of view, we need to store them somewhere, and I did it by storing to maps, but if you have any other idea please make a suggestion. Also about this `logoutJmxUser`, I will try to make changes in `logout` method if it is possible.

Reply:
As we need to provide information which user to logout, we have two proposals: 
1. Write new overloaded `logout` function to receive one parameter (`logout(Subject currentUser)`) and use it for logout in JmxShiroAuthenticator. Still need to add new function to SecurityService api.

2. Modified existing logout function to receive one parameter, and when call it use _null_ to behavior like before, and in cases like this with jmx use Subject provided from method where it is called. 
`public void logout(Subject currentUser) { 
    if (currentUser == null) {
      currentUser = getSubject();
    }
    try {
      logger.debug("Logging out " + currentUser.getPrincipal());
      currentUser.logout();
    } catch (ShiroException e) { ...`

What is your opinion? Do you have any other suggestion?